Ddumesnil/drupal

Dillon-Dumesnil opened the pull request:
@edx/pipeline-team Please review. Also, thoughts on making the deploy task into a pattern or just writing the clear caches script into the clear caches jobs directly in the deploy_marketing_site file? I use both the same number of times, but once as a pattern and once inline.
@edx/ecommerce FYI. Clinton, please review.
https://openedx.atlassian.net/browse/ECOM-4918
https://openedx.atlassian.net/browse/ECOM-4921
https://openedx.atlassian.net/browse/ECOM-4924

schenedx commented:
@Dillon-Dumesnil Can you put the jira ticket to the description of this PR?

macdiesel commented:
@Dillon-Dumesnil is this deployed to the sandbox gocd server yet? I would like to take a look at it in the UI as well.

macdiesel commented on edxpipelines/patterns/tasks.py:
```diff
@@ -313,6 +313,48 @@ def _fetch_secure_repo(job, secure_dir, secure_repo_envvar, secure_version_envva
     )
 
 
+def generate_fetch_secure_repository(job, repo_dir, runif="passed"):
```
@Dillon-Dumesnil I don't think you actually call this from your scripts and believe this code was removed. Have you rebased?

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,173 @@
+        {
+            'PRIVATE_GITHUB_KEY': config['private_github_key'],
+            'PRIVATE_MARKETING_REPOSITORY_URL': config['mktg_repository_url'],  # I dunno, you pick the name, mktg_repository_url will go in a <config>.yml file that you feed this script
+            'PRIVATE_MARKETING_REPOSITORY_VERSION': config['mktg_repository_version'], # this can be overridden by the person running the pipeline when they click play + on the UI
```
Does not need to be encrypted.

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,173 @@
+            'PRIVATE_MARKETING_REPOSITORY_URL': config['mktg_repository_url'],  # I dunno, you pick the name, mktg_repository_url will go in a <config>.yml file that you feed this script
+            'PRIVATE_MARKETING_REPOSITORY_VERSION': config['mktg_repository_version'], # this can be overridden by the person running the pipeline when they click play + on the UI
+            'PRIVATE_ECOM_REPOSITORY_URL': config['ecom_repository_url'],
+            'PRIVATE_ECOM_REPOSITORY_VERSION': config['ecom_repository_version'],
```
does not need to be encrypted

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,173 @@
+            'PRIVATE_ECOM_REPOSITORY_URL': config['ecom_repository_url'],
+            'PRIVATE_ECOM_REPOSITORY_VERSION': config['ecom_repository_version'],
+            # Any further variables you need inside this stage to push to aquia should go here, you'll also need the url for the aquia remote
+            'PRIVATE_ACQUIA_REMOTE': config['acquia_remote_url'],
```
Are there secrets in the url?

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,173 @@
+    )
+
+    tasks.generate_target_directory(push_to_drupal_job)
+    tasks.fetch_gomatic_secure(push_to_drupal_job, 'gomatic-secure')
```
Why do you need to fetch gomatic-secure?

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,173 @@
+                            'PRIVATE_MARKETING_REPOSITORY_VERSION', 'edx-mktg')
+
+    # Checkout and pull the master branch of marketing repo.
+    push_to_drupal_job.add_task(
```
seems redundant, shouldn't _fetch_secure_repo be enough? https://github.com/edx/edx-gomatic/blob/4514551a4be3b2412b551310b961d750e1c09134/edxpipelines/patterns/tasks.py#L287-L287

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,173 @@
+                # affected since we don't use minutes or seconds.
+                'echo -n "release-`date +%Y-%m-%d`" > ../target/tag_name.txt && ',
+                # Not sure this is right, but since my working_dir is edx/mktg, figured I wouldn't want to make target inside of it.
+                'echo -n "`date +B\ %d,\ %Y`" > ../target/message.txt && ',
```
I wouldn't bother with putting this to a file if you are not passing it back as an artifact. Why not just execute this on line 86 instead of the cat?

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,173 @@
+            [
+                '/bin/bash',
+                '-c',
+                'TAG_NAME=`cat ../target/tag_name.txt` && ',
```
you need to fetch the tag_name.txt artifact for this stage before you can use it in this task.

Dillon-Dumesnil replied:
Done

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,173 @@
+            [
+                '/bin/bash',
+                '-c',
+                'TAG_NAME=`cat ../target/tag_name.txt` && ',
```
you need to fetch the tag_name.txt artifact for this stage before you can use it in this task for this stage.

Dillon-Dumesnil replied:
Done

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,173 @@
+    # tasks.fetch_secure_repo(push_to_drupal_job, 'ecom-secure', 'PRIVATE_ECOM_REPOSITROY_URL',
+    #                         'PRIVATE_ECOM_REPOSITORY_VERSION', 'ecom-secure')
+    tasks.generate_requirements_install(clear_stage_caches_job, 'tubular')
+    tasks.flush_drupal_caches('test', 'PRIVATE_ACQUIA_USERNAME', 'PRIVATE_ACQUIA_PASSWORD')
```
Should PRIVATE_ACQUIA_USERNAME and PRIVATE_ACQUIA_PASSWORD be bash variables (prefixed with a $)?

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,173 @@
+        .set_git_material(GitMaterial("https://github.com/edx/tubular", polling=False))
+
+    push_to_drupal_stage = pipeline.ensure_stage("github_workflow")
+    push_to_drupal_job = push_to_drupal_stage.ensure_job("push_to_drupal_job")
```
Please make all stage and job names static variables, you will need to reference them later to pull the artifacts in to other stages/jobs and I would prefer to avoid a bunch of repeating strings.

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,173 @@
+from gomatic import *
+
+import click
+import edxpipelines.utils as utils
```
Add the sys.path stuff here.

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,173 @@
+
+    # Stage to clear caches in stage
+    clear_stage_caches_stage = pipeline.ensure_stage("clear_drupal_stage_caches")
+    clear_stage_caches_job = clear_stage_caches_stage.ensure_job('clear_stage_caches_job')
```
Please make all stage and job names static variables, you will need to reference them later to pull the artifacts in to other stages/jobs and I would prefer to avoid a bunch of repeating strings.

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,173 @@
+
+    # Stage to deploy to stage
+    deploy_stage_for_stage = pipeline.ensure_stage("deploy_drupal_to_stage")
+    deploy_job_for_stage = deploy_stage_for_stage.ensure_job('deploy_to_stage_job')
```
Please make all stage and job names static variables, you will need to reference them later to pull the artifacts in to other stages/jobs and I would prefer to avoid a bunch of repeating strings.

macdiesel commented:
@Dillon-Dumesnil First pass on this complete. LMK when I can view the pipeline in the UI.

Dillon-Dumesnil commented:
Pipeline is available on gocd.sandbox.edx.org for UI viewing.
Version 1.1 has been committed addressing the comments from version 1. Mostly changing environment parameters to unencrypted secure variables and adding static names for stages and jobs. Artifacts are now pulled in at appropriate stages.

doctoryes commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,220 @@
+    #         'PRIVATE_ACQUIA_USERNAME': config['acquia_username'],
+    #         'PRIVATE_ACQUIA_PASSWORD': config['acquia_password']
+    #     }
+    # )
```
If you're not planning on doing this, then remove it. A single comment above would be better, if needed at all.

macdiesel replied:
@doctoryes I will work with @Dillon-Dumesnil to make sure that all of the env vars are using ensure_encrypted_environment_variables.

doctoryes commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,220 @@
+                '-c',
+                # Possibly unnecessary since it should be covered under the repo_version
+                # env variable, but included to double check
+                '/usr/bin/git checkout master && ',
```
This shouldn't be left in a "not sure if needed" state.

macdiesel replied:
Don't need the comma here.

macdiesel replied:
Actually, you don't need this at all, the private repo pattern will cover this. Please remove.

doctoryes commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,220 @@
+    clear_stage_caches_job = clear_stage_caches_stage.ensure_job(CLEAR_STAGE_CACHES_JOB_NAME)
+    # Do not need to use ecom-secure until we start using drush acquia commands.
+    # tasks._fetch_secure_repo(push_to_drupal_job, 'ecom-secure', 'PRIVATE_ECOM_REPOSITROY_URL',
+    #                         'ECOM_REPOSITORY_VERSION', 'ecom-secure')
```
Get rid of the code comment here - it contains a misspelling anyway.

doctoryes commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,220 @@
+    clear_prod_caches_job = clear_prod_caches_stage.ensure_job(CLEAR_PROD_CACHES_JOB_NAME)
+    # Do not need to use ecom-secure until we start using drush acquia commands.
+    # tasks._fetch_secure_repo(push_to_drupal_job, 'ecom-secure', 'PRIVATE_ECOM_REPOSITROY_URL',
+    #                         'ECOM_REPOSITORY_VERSION', 'ecom-secure')
```
Ditto here.

doctoryes commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,220 @@
+                '--password $PRIVATE_ACQUIA_PASSWORD '
+                '--tag $TAG_NAME'
+            ],
+            working_dir='tubular/'
```
Nit: For all these `working_dir`s - is the trailing slash really needed? IIRC, it isn't - but I could be wrong. It'd be better to leave it off.

macdiesel replied:
@doctoryes is correct.

macdiesel commented on edxpipelines/patterns/tasks.py:
```diff
@@ -445,3 +445,39 @@ def generate_run_app_playbook(job, secure_dir, playbook_path, runif="passed", **
+            [
+                'bin/bash',
+                '-c',
+                'drush -y @edx.{site_env} cc all && ',
```
@Dillon-Dumesnil I'm not sure this will work as expected. I do not think that the string replacement will work here as this string is not concatenated with the strings below. Because of the comma this string is on it's own in the array.

Also I'm not sure this task will execute correctly again due to the comma I'm not sure that the go-agent to bash translation will like this. Is it executing as expected on the server?

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,220 @@
+            'PRIVATE_ECOM_REPOSITORY_URL': config['ecom_repository_url'],
+            'PRIVATE_ACQUIA_REMOTE': config['acquia_remote_url'],
+            # Unless we start using drush commands again, Acquia account credentials can be stored in config
+            # since we need to directly pass in its contents as opposed to drush searching for the file.
```
I'm a bit confused by this comment. Since you just put all that work in to tubular I doubt we will be migrating to drush. This comment can be removed.

Question: Are there any other things related to drush in this PR that will not be used? (comments, unused config, commented out code). 
Please remove all of them.

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,220 @@
+        HostRestClient(config['gocd_url'], config['gocd_username'], config['gocd_password'], ssl=True))
+
+    pipeline = configurator \
+        .ensure_pipeline_group("marketing_deployment") \
```
This pipeline should go in the "Deploy" group.

Dillon-Dumesnil replied:
@macdiesel Can you clarify this comment? Do you just mean switch "marketing_deployment" to something more like "deploy-marketing-site"?

macdiesel replied:
```
pipeline = configurator \
        .ensure_pipeline_group("Deploy") \
```

Dillon-Dumesnil replied:
@macdiesel Does .ensure_replacement_of_pipeline need to change to "Deploy" too? 
I'm curious how I end up naming it something like "deploy-marketing-site" if I'm using .ensure_pipeline_group("Deploy")

macdiesel commented:
Also mixing of ' and ", pick one. It looks like the majority of the repository uses '

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,220 @@
+
+    tasks.generate_requirements_install(push_to_acquia_job, 'tubular')
+    tasks.generate_target_directory(push_to_acquia_job)
+    tasks._fetch_secure_repo(push_to_acquia_job, 'edx-mktg', 'PRIVATE_MARKETING_REPOSITORY_URL',
```
@doctoryes what were you thinking the pattern would be for this "private" method? Should it be called directly (if so we should remove the _ prefix)? Or should a new pattern be made called something like fetch_edx_mktg in the patterns.tasks module?

doctoryes replied:
A new pattern should be made. The `_` prefix was to discourage anyone from calling it.

Dillon-Dumesnil replied:
@macdiesel @doctoryes So this comes back to where I made that pattern of fetch_secure_repo which just calls _fetch_secure_repo, but takes in the same arguments. Should I revert back to that or make one specifically for edx-mktg? I was under the assumption that we did not want to hard-code these in and instead have a generic method for all secure repos.

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,220 @@
+                # Writing dates to a file should help with any issues dealing with a job
+                # taking place over two days (23:59:59 -> 00:00:00). Only the day can be
+                # affected since we don't use minutes or seconds.
+                'echo -n "release-`date +%Y-%m-%d`" > ../target/tag_name.txt && ',
```
don't

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,220 @@
+                # taking place over two days (23:59:59 -> 00:00:00). Only the day can be
+                # affected since we don't use minutes or seconds.
+                'echo -n "release-`date +%Y-%m-%d`" > ../target/tag_name.txt && ',
+                'TAG_NAME=`cat ../target/tag_name.txt` && ',
```
need

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,220 @@
+                # affected since we don't use minutes or seconds.
+                'echo -n "release-`date +%Y-%m-%d`" > ../target/tag_name.txt && ',
+                'TAG_NAME=`cat ../target/tag_name.txt` && ',
+                '/usr/bin/git tag -a $TAG_NAME -m "Release for `date +B\ %d,\ %Y`" && ',
```
commas

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,220 @@
+                'bin/bash',
+                '-c',
+                'TAG_NAME=`cat ../target/tag_name.txt` && ',
+                '/usr/bin/git remote add acquia $PRIVATE_ACQUIA_REMOTE && ',
```
remove commas on these 2 lines.

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,220 @@
+
+    # Stage to clear caches in prod
+    clear_prod_caches_stage = pipeline.ensure_stage(CLEAR_PROD_CACHES_STAGE_NAME)
+    # Should only run if manually started
```
The code tells us this very plainly, you can remove this comment.

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,220 @@
+
+    tasks.generate_requirements_install(deploy_job_for_prod, 'tubular')
+    tasks.generate_target_directory(deploy_job_for_prod)
+    deploy_job_for_prod.add_task(FetchArtifactTask(**artifact_params))
```
Can we give artifact_params a more descriptive name please?

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,220 @@
+            [
+                '/bin/bash',
+                '-c',
+                'TAG_NAME=`cat ../target/tag_name.txt` && ',
```
errant comma?

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,220 @@
+    }
+    deploy_job_for_stage.add_task(FetchArtifactTask(**artifact_params))
+
+    deploy_job_for_stage.add_task(
```
This stage is written twice with the only difference being the environment, please make it a pattern.

macdiesel commented:
@Dillon-Dumesnil I have finished my first pass.

macdiesel commented on edxpipelines/patterns/tasks.py:
```diff
@@ -445,3 +445,39 @@ def generate_run_app_playbook(job, secure_dir, playbook_path, runif="passed", **
+    """
+    Flushes all drupal caches (including the varnish cache)
+    Assumes the drupal root is located in edx-mktg/docroot. If changed, change the working dir.
+
```
Please add doc that states that 2 environment variables must be defined for this task the username and password

macdiesel commented on edxpipelines/patterns/tasks.py:
```diff
@@ -445,3 +445,39 @@ def generate_run_app_playbook(job, secure_dir, playbook_path, runif="passed", **
+        job (gomatic.job.Job): the gomatic job to which the task will be added
+        site_env (str): The environment to clear caches from. Choose 'test' for stage and 'prod' for prod
+        username (str): The Acquia username used to run the command.
+        password (str): The Acquia password used to run the command.
```
call out here that these should be environment variables setup in the pipeline and that the caller is passing in the name of that variable.

Dillon-Dumesnil commented:
@edx/pipeline-team All comments have been responded to. Please review updated version

macdiesel commented on edxpipelines/patterns/tasks.py:
```diff
@@ -445,3 +469,106 @@ def generate_run_app_playbook(job, secure_dir, playbook_path, runif="passed", **
+            [
+                '/bin/bash',
+                '-c',
+                'TAG_NAME=`cat ../target/tag_name.txt` && '
```
have you tested this to make sure that TAG_NAME is available to the next command? I'm not sure it will work unless you export it. Also change this && to ; and you don't need the space at the end.

Dillon-Dumesnil replied:
I haven't been able to test it yet because my pipeline is unable to work on sandbox and I haven't tried putting it on tools yet.
I would argue it should be && because calling the script will fail if the tag is not correct. I want to ensure that process succeeds before starting the script.

macdiesel replied:
Better:
```
               '/usr/bin/python',
                'scripts/drupal_deploy.py',
                '--env ${site_env}'.format(site_env=site_env),
                '--username $PRIVATE_ACQUIA_USERNAME',
                '--password $PRIVATE_ACQUIA_PASSWORD',
                '--tag `cat ../target/tag_name.txt`'
```

Dillon-Dumesnil replied:
I like that. Will update

cpennington replied:
Nit: the `$(...)` syntax for subprocessing in bash is often clearer, and nests properly.

Dillon-Dumesnil replied:
Do I not need spaces at the end of every line? Looking at the patterns generate_run_migrations made me think I did. When do you need spaces vs. when don't you?

macdiesel replied:
when these items are executing in bash using && to chain commands together, gocd prefers everything to be on the same line, hence the spaces and no commas, this concatenates the string.

When you don't need to chain commands together, you should call the binary executable and pass all the commands as an array. When you do this GoCD performs extra formatting on the command to ensure the correct spaces are in place between each item (parameter, argument, etc) in the array.

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,177 @@
+                # taking place over two days (23:59:59 -> 00:00:00). Only the day can be
+                # affected since we don't use minutes or seconds.
+                'echo -n "release-`date +%Y-%m-%d`" > ../target/tag_name.txt && '
+                'TAG_NAME=`cat ../target/tag_name.txt` && '
```
replace with $(cat ../target/tag_name.txt)

macdiesel commented on edxpipelines/pipelines/deploy_marketing_site.py:
```diff
@@ -0,0 +1,177 @@
+            [
+                'bin/bash',
+                '-c',
+                'TAG_NAME=`cat ../target/tag_name.txt` && '
```
replace with $(cat ../target/tag_name.txt)

macdiesel commented on edxpipelines/patterns/tasks.py:
```diff
@@ -445,3 +469,106 @@ def generate_run_app_playbook(job, secure_dir, playbook_path, runif="passed", **
+    )
+
+
+def flush_drupal_caches(job, site_env):
```
generate_flush_drupal_caches

macdiesel commented on edxpipelines/patterns/tasks.py:
```diff
@@ -445,3 +469,106 @@ def generate_run_app_playbook(job, secure_dir, playbook_path, runif="passed", **
+    )
+
+
+def drupal_deploy(job, site_env):
```
generate_drupal_deploy

macdiesel commented:
@Dillon-Dumesnil pass complete.

Dillon-Dumesnil commented:
@edx/pipeline-team Updated version with responses to comments

macdiesel commented on edxpipelines/patterns/tasks.py:
```diff
@@ -446,3 +470,117 @@ def generate_run_app_playbook(job, secure_dir, playbook_path, runif="passed", **
+                '--env', '{site_env}'.format(site_env=site_env),
+                '--username', '$PRIVATE_ACQUIA_USERNAME',
+                '--password', '$PRIVATE_ACQUIA_PASSWORD',
+                '--tag', '$(cat ../target/tag_name.txt)'
```
Newline after --tag',

macdiesel commented:
One small thing and add a the script to the whitelist in config.yml. Set enabled to False and then we are 👍

Dillon-Dumesnil commented:
Except for the part where I am adding in the revert code. :( Will ping you once the commit is better

macdiesel commented:
@Dillon-Dumesnil this PR is getting large, I would like to pull it in with the config disabled. Any additional changes for the revert Pipeline can come in a new PR. This will make it easier to review.

macdiesel commented on config.yml:
```diff
@@ -41,9 +41,21 @@ tools:
 
+  - script: edxpipelines/pipelines/deploy_marketing_site.py
+    input_files:
+      - &tools-admin "../gomatic-secure/gocd/vars/tools.admin.yml"
```
this should be simply:
```
*tools-admin
```

macdiesel commented on config.yml:
```diff
@@ -41,9 +41,21 @@ tools:
+
+  - script: edxpipelines/pipelines/deploy_marketing_site.py
+    input_files:
+      - &sandbox-admin "../gomatic-secure/gocd/vars/sandbox.admin.yml"
```
this should be
```
*sandbox-admin
```

macdiesel commented:
👍
